### PR TITLE
[DEV-12952] Add Configurable Retry Behavior

### PR DESCRIPTION
**Pull request by mawelborn**
## Summary

This PR implements configurable, best-practices retry behavior for both `IndicoClient` and `AsyncIndicoClient` with consistent socket connect and read timeouts to replace the existing retry behavior that is only present on the async client and is not configurable. `IndicoConfig` has new options to control retry behavior, with reasonable defaults:

```python
IndicoConfig(
    retry_count: int = 4,  # Retry API calls this many times.
    retry_wait: float = 1,  # Wait this many seconds after the first error before retrying.
    retry_backoff: float = 4,  # Multiply the wait time by this amount for each additional error.
    retry_jitter: float = 0.5,  # Add a random amount of time (up to this percent as a decimal) to the wait time to prevent simultaneous retries.
)
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] Jira story/task is put into PR title
- [x] Code has been self-reviewed
- [x] Hard-to-understand areas have been commented
- [x] Documentation has been updated
- [x] My changes generate no new warnings
- [x] Unit tests have been added for base functionality and known edge cases
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by mawelborn** [comment]
Hi @IndicoDataSolutions/pr-be-indicodata-ai folks! I rebased this PR onto main and have all the checks passing. Would appreciate a review when you have a chance. This functionality will be super useful for ProServe and customers.

File: indico/config/config.py [hunk 1]
```diff
@@ -42,6 +46,11 @@ def __init__(self, **kwargs: "Any"):
         self.requests_params: "Optional[AnyDict]" = None
         self._disable_cookie_domain: bool = False
 
+        self.retry_count: int = int(os.getenv("INDICO_retry_count", "4"))
```
Comment (thearchitector) [inline indico/config/config.py]: caps?

**Comment by mawelborn** [comment]
Hello again! Sorry, just now getting back to this.

I don't have strong opinions regarding the retry implementation--whether we ship our own or use a library like Tenacity. I used the retry implementation from the toolkit since it's a known quantity, and shipping our own was already being done for AsyncIndicoClient.

Do I need to swap out the implementation or make any other changes? Or is the current implementation good enough for now?

Thanks!

File: indico/http/client.py [hunk 2]
```diff
@@ -232,6 +241,14 @@ def __init__(self, config: "Optional[IndicoConfig]" = None):
...
         self.config = config or IndicoConfig()
         self.base_url = f"{self.config.protocol}://{self.config.host}"
+        self._decorate_with_retry = retry(
+            aiohttp.ClientConnectionError,
+            count=self.config.retry_count,
+            wait=self.config.retry_wait,
+            backoff=self.config.retry_backoff,
+            jitter=self.config.retry_jitter,
+        )
+        self._make_request = self._decorate_with_retry(self._make_request)  # type: ignore[method-assign]
```
[the remark this answers is not in the document]
Comment (thearchitector) [reply indico/http/client.py]: ah yeah, i guess i was thinking more of [this pattern](https://tenacity.readthedocs.io/en/latest/index.html#retrying-code-block)

but this works too!

> Callers can change the retry behavior of the next call of an existing client if needed.
wonder too if it would be useful to expose per-request retry overrides. like instead of a multiple separate request kwargs, package them together in a dataclass or something and expose `retries=Retry(...)` both at the client init level and the request level

**Review by thearchitector — APPROVED** [review]
